include/RAJA/exec-openmp/scan_openmp.hxx

```diff
@@ -0,0 +1,169 @@
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/config.hxx"
+
+#include "RAJA/exec-sequential/scan_sequential.hxx"
+
+#include <omp.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+namespace RAJA
+{
+
+RAJA_INLINE
+int firstIndex(int n, int p, int pid)
+{
+  return static_cast<size_t>(n * pid) / p;
+}
+
+template <typename Iter, typename BinFn, typename Value>
+void inclusive_scan_inplace(omp_parallel_for_exec,
+                            Iter begin,
+                            Iter end,
+                            BinFn f,
+                            Value v)
+{
+  int n = end - begin;
+  int p = omp_get_max_threads();
+  std::vector<Value> sums(p, v);
+#pragma omp parallel
+  {
+#pragma omp for schedule(static)
+    for (int pid = 0; pid < p; ++pid) {
+      const int i0 = firstIndex(n, p, pid);
+      const int i1 = firstIndex(n, p, pid + 1);
+      inclusive_scan_inplace(seq_exec{}, begin + i0, begin + i1, f, v);
+      sums[pid] = *(begin + i1 - 1);
+    }
+
+#pragma omp single
+    exclusive_scan_inplace(seq_exec{}, sums.begin(), sums.end(), f, v);
+
+#pragma omp for schedule(static)
+    for (int pid = 0; pid < p; ++pid) {
+      const int i0 = firstIndex(n, p, pid);
+      const int i1 = firstIndex(n, p, pid + 1);
+      for (int i = i0; i < i1; ++i) {
+        begin[i] = f(*(begin + i), sums[pid]);
+      }
+    }
+  }
+}
+
+template <typename Iter, typename BinFn, typename Value>
+void exclusive_scan_inplace(omp_parallel_for_exec,
+                            Iter begin,
+                            Iter end,
+                            BinFn f,
+                            Value v)
+{
+  int n = end - begin;
+  int p = omp_get_num_threads();
+  std::vector<Value> sums(p, v);
+#pragma omp parallel
+  {
+#pragma omp for schedule(static)
+    for (int pid = 0; pid < p; ++pid) {
+      const int i0 = firstIndex(n, p, pid);
+      const int i1 = firstIndex(n, p, pid + 1);
+      exclusive_scan_inplace(seq_exec{}, begin + i0, begin + i1, f, v);
+      sums[pid] = begin[i1 - 1];
+    }
+
+#pragma omp single
+    exclusive_scan_inplace(seq_exec{}, sums.begin(), sums.end(), f, v);
+
+#pragma omp for schedule(static)
+    for (int pid = 0; pid < p; ++pid) {
+      const int i0 = firstIndex(n, p, pid);
+      const int i1 = firstIndex(n, p, pid + 1);
+      for (int i = i0; i < i1; ++i) {
+        *(begin + i) = f(*(begin + i), sums[pid]);
+      }
+    }
+  }
+}
+
+template <typename Iter, typename OutIter, typename BinFn, typename Value>
+void inclusive_scan(omp_parallel_for_exec exec,
+                    Iter begin,
+                    Iter end,
+                    OutIter out,
+                    BinFn f,
+                    Value v)
+{
+  std::copy(begin, end, out);
```

Comment:
lazily copy and execute in-place. O(n) additional memory reads/writes; however, most likely optimized with a memcpy. Okay for initial version. Will improve with future enhancement